reactor-core/src/main/java/reactor/core/publisher/FluxSwitchOnFirst.java

```diff
@@ -655,4 +501,107 @@ boolean drainRegular() {
             }
         }
     }
+
+    static final class SwitchOnFirstInnerSubscriber<T> implements InnerConsumer<T> {
+
+        final AbstractSwitchOnFirstInner<?, T> parent;
+        final CoreSubscriber<? super T> inner;
+
+        SwitchOnFirstInnerSubscriber(
+                AbstractSwitchOnFirstInner<?, T> parent,
+                CoreSubscriber<? super T> inner) {
+            this.parent = parent;
+            this.inner = inner;
+        }
+
+        @Override
+        public Context currentContext() {
+            return inner.currentContext();
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            inner.onSubscribe(s);
+        }
+
+        @Override
+        public void onNext(T t) {
+            inner.onNext(t);
+        }
+
+        @Override
+        public void onError(Throwable throwable) {
+            parent.cancel();
+            inner.onError(throwable);
+        }
+
+        @Override
+        public void onComplete() {
+            parent.cancel();
+            inner.onComplete();
+        }
+
+        @Override
+        public Object scanUnsafe(Attr key) {
+            if (key == Attr.PARENT) return parent;
+            if (key == Attr.ACTUAL) return inner;
+
+            return null;
+        }
+    }
+
+    static final class SwitchOnFirstConditionalInnerSubscriber<T> implements InnerConsumer<T>,
+                                                                             Fuseable.ConditionalSubscriber<T> {
+
+        final AbstractSwitchOnFirstInner<?, ? super T>  parent;
+        final Fuseable.ConditionalSubscriber<? super T> inner;
+
+        SwitchOnFirstConditionalInnerSubscriber(
+            AbstractSwitchOnFirstInner<?, ? super T> parent,
+            Fuseable.ConditionalSubscriber<? super T> inner
+        ) {
+            this.parent = parent;
+            this.inner = inner;
+        }
+
+        @Override
+        public Context currentContext() {
+            return inner.currentContext();
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            inner.onSubscribe(s);
+        }
+
+        @Override
+        public void onNext(T t) {
+            inner.onNext(t);
+        }
+
+        @Override
+        public boolean tryOnNext(T t) {
+            return inner.tryOnNext(t);
+        }
+
+        @Override
+        public void onError(Throwable throwable) {
+            parent.cancel();
```

Comment:
again, should this guard against systematic call to `parent#cancel()` if the parent is already `done`?

Reply:
fixed